todo_app/app.py

```diff
@@ -1,26 +1,77 @@
 from flask import Flask, redirect, render_template, request
-from todo_app.data.mongo_items import add_card, get_cards, get_lists, move_card, read_env_deatils
+from todo_app.data.mongo_items import add_card, get_cards, get_lists, move_card, read_env_deatils, get_myrole, get_currentuser, role_required
 from todo_app.flask_config import Config
 from todo_app.View_Class import ViewModel
+from flask_login import LoginManager, login_required, login_user
+from os import getenv
+import requests
+from todo_app.data.user_class import User
+from datetime import timedelta
 
 def create_app():
     read_env_deatils()
     app = Flask(__name__)
     app.config.from_object(Config())
+    login_manager = LoginManager()
+    CLIENTID = getenv('CLIENTID')
+    CLIENTSECRET = getenv('CLIENTSECRET')
+    BASEURL = getenv('URL')
+    STATE = getenv('STATE')
+    app.config['LOGIN_DISABLED'] = getenv('LOGIN_DISABLED') == 'True'
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        url = f"https://github.com/login/oauth/authorize?client_id={CLIENTID}&state={STATE}&redirect_uri={BASEURL}/login/callback&allow_signup=true"
+        return redirect(url)
+    
+    @login_manager.user_loader
+    def load_user(user_id):
+        user_name = user_id.split(",")[0].split(":")[1].replace("'","")
+        user_id = user_id.split(",")[1].split(":")[1].replace("'","")
+        thisuser = User(user_id,user_name)
+        return (thisuser)
+    login_manager.init_app(app)
 
     @app.route('/')
+    @login_required
     def index():
         item_view_model = ViewModel(get_cards())
-        return render_template('index.html', view_model=item_view_model, lists=get_lists())
+        ThisUser = get_currentuser()
+        return render_template('index.html', view_model=item_view_model, lists=get_lists(), user1=ThisUser, role=get_myrole(ThisUser.id))
 
     @app.route('/new', methods=['POST'])
+    @login_required
+    #@role_required('writer')
     def new():
+        if role_required('writer') :
+            exit
         add_card(request.form.get('title'))
         return redirect('/')
 
     @app.route('/move/<cardID>/<newList>')
+    @login_required
+    #@role_required('writer')
     def move(cardID, newList):
+        if role_required('writer') :
+            exit
         move_card(cardID, newList)
         return redirect('/')
 
+    @app.route('/login/callback')
+    def callback():
+        recivedCode = request.args.get('code')
```

Comment:
It's optional here, but I can see you've included state above so I want to talk about that briefly! 

As you said, state is a nice tool to make cross site attacks harder. That said, we have to implement it slightly differently from here - because currently I (and any user) can see the state as part of the redirect URL, so it's immediately leaked when you have a single user! In order to work round that, the usual pattern is:

* When a user appears (e.g. in your `unauthenticated` function), we _generate_ a fresh random string for that user - this will be our state string _for that one person_
* We store that in the user's session cookie, and then send them off to GitHub with their own state
* When they come back from GitHub (i.e. here in callback) we check that the state that they came back with matches the one we can see in their session cookie
  * If it doesn't, it suggests they were redirected from GitHub that wasn't us, so we stop and warn them!
  * If it does, no issues, carry on

Reply:
Addressesd the state